engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/UUIDValidator.java

```diff
@@ -0,0 +1,174 @@
+ */
+public class UUIDValidator implements ConstraintValidator<UUID, CharSequence> {
+
+	private static final int[] GROUP_LENGTHS = { 8, 4, 4, 4, 12 };
+
+	private boolean allowEmpty;
+
+	private boolean allowNil;
+
+	private int[] version;
+
+	private int[] variant;
+
+	private LetterCase letterCase;
+
+	@Override
+	public void initialize(UUID constraintAnnotation) {
+		allowEmpty = constraintAnnotation.allowEmpty();
+		allowNil = constraintAnnotation.allowNil();
+		version = checkAndSortMultiOptionParameter( constraintAnnotation.version(), "version", 1, 15 );
+		variant = checkAndSortMultiOptionParameter( constraintAnnotation.variant(), "variant", 0, 2 );
+		letterCase = constraintAnnotation.letterCase();
+		Contracts.assertNotNull( letterCase, MESSAGES.parameterMustNotBeNull( "letterCase" ) );
+	}
+
+	private static int[] checkAndSortMultiOptionParameter(
+			int[] values,
+			String parameterName,
+			int minimum,
+			int maximum) {
+		Contracts.assertNotNull( values, MESSAGES.parameterMustNotBeNull( parameterName ) );
+		Contracts.assertNotEmpty( values, MESSAGES.parameterMustNotBeEmpty( parameterName ) );
+		for ( int value : values ) {
+			Contracts.assertTrue(
+					value >= minimum,
+					MESSAGES.parameterShouldBeGreaterThanOrEqualTo( parameterName, minimum )
+			);
+			Contracts.assertTrue(
+					value <= maximum,
+					MESSAGES.parameterShouldBeLessThanOrEqualTo( parameterName, maximum )
+			);
+		}
+		Arrays.sort( values );
+		return values;
+	}
+
+	@Override
+	public boolean isValid(CharSequence value, ConstraintValidatorContext context) {
+		if ( value == null ) {
+			return true;
+		}
+		int valueLength = value.length();
+		if ( valueLength == 0 ) {
+			return allowEmpty;
+		}
+		else if ( valueLength != 36 ) {
+			return false;
+		}
+
+		int groupIndex = 0;
+		int groupLength = 0;
+		int checksum = 0;
+		int version = -1;
+		int variant = -1;
+		for ( int charIndex = 0; charIndex < valueLength; charIndex++ ) {
+
+			char ch = value.charAt( charIndex );
+
+			if ( ch == '-' ) {
+				groupIndex++;
+				groupLength = 0;
+			}
+			else {
+
+				groupLength++;
+				if ( groupLength > GROUP_LENGTHS[groupIndex] ) {
+					return false;
+				}
+
+				int numericValue = Character.digit( ch, 16 );
+				if ( numericValue == -1 ) {
+					// not a hex digit
+					return false;
+				}
+				if ( letterCase == LetterCase.LOWER_CASE && numericValue > 9 && !Character.isLowerCase( ch ) ) {
+					return false;
+				}
+				if ( letterCase == LetterCase.UPPER_CASE && numericValue > 9 && !Character.isUpperCase( ch ) ) {
+					return false;
+				}
+				checksum += numericValue;
+				version = extractVersion( version, charIndex, numericValue );
+				variant = extractVariant( variant, charIndex, numericValue );
+
+			}
+
+		}
+
+		if ( checksum == 0 ) {
+			return allowNil;
+		}
+		else {
+			if ( Arrays.binarySearch( this.version, version ) == -1 ) {
+				return false;
+			}
+			return Arrays.binarySearch( this.variant, variant ) != -1;
+		}
+
```

Comment:
Shouldn't there be some sort of check in here that the version belongs to {1,2,3,4,5} ?
In the UUIDValidatorTest you do verify that versions match, but you're using versions that are (IIUC) are not allowed?

Reply:
I hope that I understand you correctly: The version can have a value that is not yet officially existing. But maybe in the future there will be a new kind of UUID and the implementation will support it.

Reply:
Ok, in general I'm somewhat more in favor of not foreseeing everything that could potentially happen in the future, but it's not a big issue, I'll leave it up to you. I do like your validation implementation, I think it's quite elegant.

Reply:
Thank you! Really cheers me up. ☺️